src/main/scala/wdlTools/eval/stdlib/BaseStdlib.scala

```diff
@@ -1,197 +1,227 @@
+      }
+      .toVector
+    if (lines.size < 2) {
+      throw new EvalException(s"read_object : file ${file.toString} must contain at least two",
                               text,
                               docSourceUrl)
-    val keys = lines(0).split("\t")
-    val values = lines(1).split("\t")
-    if (keys.size != values.size)
+    }
+    val keys = lines.head
+    V_Array(lines.tail.map(values => kvToObject(keys, values, text)))
+  }
+
+  private def kvToObject(keys: Vector[String],
+                         values: Vector[String],
+                         text: TextSource): V_Object = {
+    if (keys.size != values.size) {
       throw new EvalException(
-          s"read_object : the number of keys (${keys.size}) must be the same as the number of values (${values.size})",
+          s"""read_object : the number of keys (${keys.size}) 
+             |must be the same as the number of values (${values.size})""".stripMargin
+            .replace("\t", " "),
           text,
           docSourceUrl
       )
-
-    // Note all the values are going to be strings here. This probably isn't what
-    // the user wants.
+    }
+    // Note all the values are going to be strings here. This probably isn't what the user wants.
     val m = (keys zip values).map {
       case (k, v) =>
         k -> V_String(v)
     }.toMap
     V_Object(m)
   }
 
-  // Array[Object] read_objects(String|File)
-  private def read_objects(args: Vector[V], text: TextSource): V_Array = {
-    val file = getWdlFile(args, text)
-    val content = iosp.readFile(file.value)
-    val lines = content.split("\n")
-    if (lines.size < 2)
-      throw new EvalException(s"read_object : file ${file.toString} must contain at least two",
-                              text,
-                              docSourceUrl)
-    val keys = lines.head.split("\t")
-    val objects = lines.tail.map { line =>
-      val values = line.split("\t")
-      if (keys.size != values.size)
-        throw new EvalException(
-            s"read_object : the number of keys (${keys.size}) must be the same as the number of values (${values.size})",
-            text,
-            docSourceUrl
-        )
-      // Note all the values are going to be strings here. This probably isn't what
-      // the user wants.
-      val m = (keys zip values).map {
-        case (k, v) =>
-          k -> V_String(v)
-      }.toMap
-      V_Object(m)
-    }.toVector
-    V_Array(objects)
-  }
-
   // mixed read_json(String|File)
-  private def read_json(args: Vector[V], text: TextSource): V = {
+  //
+  // since: draft-1
+  // deprecation: beginning in draft-2, URI parameter is not supported
+  protected def read_json(args: Vector[V], text: TextSource): V = {
     val file = getWdlFile(args, text)
-    val content = iosp.readFile(file.value)
+    val content = iosp.readFile(file.value, text)
     try {
-      Serialize.fromJson(content.parseJson)
+      Serialize.fromJson(content.mkString.parseJson)
```

Comment:
Here and elsewhere, you are reading a file, splitting into lines, and then concatenating. I don't know how that will play out with binary files.

Reply:
Source.fromFile reads and splits the file in a streaming fashion, i.e. it detects newlines as it reads the file.

The larger question is whether the WDL I/O functions are expected to handle binary files. To me it is pretty apparent that they are not. WDL has only String type, not Bytes, so even if you did read in binary data there would be no way to use it in a WDL script. I think we can safely add the explicit limitation to wdlTools that only text files are supported by the read_* functions, and the behavior is undefined if the user tries to read a binary file.

Another question is what character encoding should be used to read text files. Again, the spec does not say. There are two reasonable options - UTF8 or the default encoding of the system where the WDL is executed. For now I've gone with the latter, but made it configurable via an option in EvalConfig, with the default being `Codec.default.name`.

Reply:
I looked into this more and I think you're right - io.Source has some issues. I've reverted to using the Java nio functions for reading files.